### PR TITLE
Feature: Add and set a property to mark local authors as deleted.

### PR DESCRIPTION
**Pull request by twstokes**
This is one PR in a series towards implementing #9015.

**Current behavior:** Authors who are deleted from a blog continue to accumulate in the local store of authors in the app.

**What this change does:** Authors who are deleted from a blog continue to accumulate in the local store of authors in the app, but have the property `deletedFromBlog` set to `true`. All others are set to `false`.

**Why this change was needed:** This will prevent users from seeing deleted authors in the list to choose from once #9015 is implemented.

### To test:

1. Prepare a WordPress.com or self-hosted WordPress blog with multiple users.
2. Place a breakpoint on `BlogService+BlogAuthors.swift:36` and load the "My Site" view for the blog.
3. When the breakpoint is caught, observe the output of: `po remoteUsers` and `po blog.authors`.
- The `remoteUsers` array should contain all of the blog's remote users.
- The `blog.authors` set should match number of remote users, and the `deletedFromBlog` property should be `0` for each.
4. Continue execution of the app.
5. Delete one of the users from the blog using the WordPress Dashboard.
6. Restart WPiOS or choose "Blog Posts" and press "Back" to return to the "My Site" view.
7. When the breakpoint is caught, observe the output of the commands from step 3.
- The `remoteUsers` array should contain one less remote user (the user that was deleted is no longer in that array).
- The `blog.authors` set should still match the number of remote users from step 3, but the `deletedFromBlog` property should equal `1` for the deleted author.

### 1. Potential unintended areas of impact

- None identified.

### 2. What I did to test those areas of impact (or what existing automated tests I relied on)

**Metadata for a deleted author (Display Name, Avatar icon) should still appear when viewing the History of a Post.**

1. Make changes to a Post as multiple users (e.g. User A and User B have both made updates to the same Post).
2. With the Post presented in the app, tap the horizontal ellipses (More Options) icon at the top right.
3. Select "History".
4. Observe the display names and avatars next to each revision.
5. Remove an author from the blog.
6. Restart the app.
7. View the history for the same Post.
8. There should be no changes or regressions.

### 3. What automated tests I added (or what prevented me from doing so)

- Added `BlogServiceAuthorTests`.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.
- [x] I have documented the migration.

### CODE REVIEW

File: WordPress/WordPress.xcodeproj/project.pbxproj
Comment (twstokes) [inline WordPress/WordPress.xcodeproj/project.pbxproj]: I don't know what these changes mean, but they occurred when the data model was changed. Confirmation that this is expected would be great!

**Comment by guarani** [comment]
> 	Warnings
> ⚠️ 	PR has more than 500 lines of code changing. Consider splitting into smaller PRs if possible.
> 
> Generated by 🚫 [dangerJS](https://danger.systems/js)

This warning is unavoidable for Core Data migrations.

**Comment by guarani** [comment]
Since this involves a core data migration, are there any extra tests you recommend for testing?

**Review by guarani — APPROVED** [review]
Tested with a WP.com site and a self-hosted site and things working as expected, nice work!

**Comment by twstokes** [comment]
> Since this involves a core data migration, are there any extra tests you recommend for testing?

I put some thought into this and the Lightweight Core Data migration felt safe since we're adding a non-optional default property, and currently no other code paths reference it other than the ones in this PR.

Would instructions on installing the current version of the app and then upgrading to the version in this PR suffice, or should we go further and add an automated migration test?

Thanks for bringing this up, I'm all ears for guidance!

**Comment by guarani** [comment]
> I put some thought into this and the Lightweight Core Data migration felt safe since we're adding a non-optional default property, and currently no other code paths reference it other than the ones in this PR.

Yeah, if it was a non-optional without a default value I'd be more cautious, but this feels pretty safe.

> Would instructions on installing the current version of the app and then upgrading to the version in this PR suffice, or should we go further and add an automated migration test?

I agree this feels like it should work out of the box. The test you mention here of installing this on top of the current `develop` version is something I often do as a confidence check. I did that and it worked, so I think this can be merged.

---

Is the merge button enabled for you? If not, I'll merge it now.

**Comment by twstokes** [comment]
> Is the merge button enabled for you? If not, I'll merge it now.

Sounds great, thanks Paul for your insight! I don't get the nice button to push, not authorized.

**Comment by twstokes** [comment]
@guarani I'm going to apply https://github.com/wordpress-mobile/WordPress-iOS/pull/16351 to this once it gets merged in, to clean up these changes.

File: WordPress/WordPress.xcodeproj/project.pbxproj
Comment (twstokes) [reply WordPress/WordPress.xcodeproj/project.pbxproj]: Turns out these came from Xcode, unrelated to the migration. See: https://github.com/wordpress-mobile/WordPress-iOS/pull/16351

**Comment by guarani** [comment]
Thanks for the update! I gave this a quick look again after the changes, re-triggered the CI, and enabled auto-merge so this should be merged soon.

**Comment by twstokes** [comment]
> Thanks for the update! I gave this a quick look again after the changes, re-triggered the CI, and enabled auto-merge so this should be merged soon.

Excellent - thank you @guarani!

**Pull request merged**
